src/value.rs

```diff
@@ -5,69 +5,231 @@
+// is constant overhead (this just compares TMRs). If those match, we know
+// the lengths and structures match, so we then compare the underlying byte
+// iterators.
+impl PartialEq for Value {
+    fn eq(&self, other: &Self) -> bool {
+        self.ty == other.ty && self.raw_byte_iter().eq(other.raw_byte_iter())
+    }
+}
+impl Eq for Value {}
+
+impl PartialOrd for Value {
+    fn partial_cmp(&self, other: &Self) -> Option<core::cmp::Ordering> {
+        Some(self.cmp(other))
+    }
+}
+impl Ord for Value {
+    fn cmp(&self, other: &Self) -> core::cmp::Ordering {
+        self.ty
+            .cmp(&other.ty)
+            .then_with(|| self.raw_byte_iter().cmp(other.raw_byte_iter()))
+    }
+}
+
+impl core::hash::Hash for Value {
+    fn hash<H: core::hash::Hasher>(&self, h: &mut H) {
+        "[simplicity value]".hash(h);
+        self.ty.hash(h);
+        for val in self.raw_byte_iter() {
+            val.hash(h);
+        }
+    }
 }
 
-impl DagLike for &'_ Value {
+impl DagLike for Value {
     type Node = Value;
 
     fn data(&self) -> &Value {
         self
     }
 
     fn as_dag_node(&self) -> Dag<Self> {
-        match &self.inner {
-            ValueInner::Unit => Dag::Nullary,
-            ValueInner::Left(child) | ValueInner::Right(child) => Dag::Unary(child),
-            ValueInner::Product(left, right) => Dag::Binary(left, right),
+        if let Some((left, right)) = self.to_product() {
+            Dag::Binary(left, right)
+        } else if let Some(left) = self.to_left() {
+            Dag::Unary(left)
+        } else if let Some(right) = self.to_right() {
+            Dag::Unary(right)
+        } else {
+            assert!(self.ty.is_unit());
+            Dag::Nullary
         }
     }
 }
 
+pub struct RawByteIter<'v> {
+    value: &'v Value,
+    yielded_bytes: usize,
+}
+
+impl Iterator for RawByteIter<'_> {
+    type Item = u8;
+    fn next(&mut self) -> Option<Self::Item> {
+        if 8 * self.yielded_bytes >= self.value.ty.bit_width() {
+            None
+        } else if self.value.bit_offset % 8 == 0 {
+            self.yielded_bytes += 1;
+
+            Some(self.value.inner[self.value.bit_offset / 8 + self.yielded_bytes - 1])
+        } else {
+            self.yielded_bytes += 1;
+
+            let ret1 = self.value.inner[self.value.bit_offset / 8 + self.yielded_bytes - 1];
+            let ret2 = self
+                .value
+                .inner
+                .get(self.value.bit_offset / 8 + self.yielded_bytes)
+                .copied()
+                .unwrap_or(0);
+            let bit_offset = self.value.bit_offset % 8;
+            Some((ret1 << bit_offset) | (ret2 >> (8 - bit_offset)))
+        }
+    }
+}
+
+pub struct PreOrderIter<'v> {
+    inner: iter::Take<BitIter<RawByteIter<'v>>>,
+}
+
+impl Iterator for PreOrderIter<'_> {
+    type Item = bool;
+    fn next(&mut self) -> Option<Self::Item> {
+        self.inner.next()
+    }
+}
```

Comment:
cbbe9b5a3e7982c5b4c85b28cc37996fbcb0c7b2: Is this type useful? AFAICS, we can simply remove it. The only method that calls it returns an opaque `impl Iterator` type.

Reply:
Yeah, lemme double-back and check on this. This was one of the first things I implemented and the PR has been almost entirely rewritten since then.

Reply:
Changed both to return concrete types instead of `impl Iterator`. If the concrete types are available then they're much nicer for the user -- they are sized, we can implement additional traits on them, they can store them in structs, etc.

Reply:
We can store opaque `impl` types in structs using generics, can't we?

In any case, now `PreOrderIter` is exposed via the API and slightly more useful. Good enough.

Reply:
> We can store opaque `impl` types in structs using generics, can't we?

I'm pretty sure not. You can try it in the playground.